### PR TITLE
Add ability for admins to create new substrate communities

### PR DESCRIPTION
**Pull request by maxwellfoley3**
## Description
In the panel that allows admins to create new offchain communities, there is now the option to add a new Substrate community. 

## How has this been tested?
Manually

### CODE REVIEW

File: client/scripts/app.ts [hunk 1]
```diff
@@ -228,12 +228,7 @@ export async function selectNode(n?: NodeInfo, deferred = false): Promise<boolea
   if (app.chain && n === app.chain.meta) {
     return;
   }
-  if ((Object.values(ChainNetwork) as any).indexOf(n.chain.network) === -1
```
Comment (zakhap) [inline client/scripts/app.ts:231]: Why is this removed?

File: client/scripts/views/modals/create_community_modal.ts [hunk 2]
```diff
@@ -134,33 +138,195 @@ const CreateCommunityModal: m.Component<IAttrs, IState> = {
...
+          defaultValue: vnode.state.telegram,
+          placeholder: 'https://t.me',
+          onChangeHandler: (v) => { vnode.state.telegram = v; },
+        }),
+        m(InputPropertyRow, {
+          title: 'Github',
+          defaultValue: vnode.state.github,
+          placeholder: 'https://github.com',
+          onChangeHandler: (v) => { vnode.state.github = v; },
+        }),
```
Comment (zakhap) [inline client/scripts/views/modals/create_community_modal.ts:243]: We're going to have to add another field, probably a larger one, for what's known as a "NodeSpec". This is a json blob that we have to validate (we are already doing so in the manage community modal where the admins can edit+update the NodeSpec, you can copy the validation schema from there). This "nodespec" is added to the `chain.ts` model on the server.

File: server/routes/addChainNode.ts [hunk 3]
```diff
@@ -48,6 +48,13 @@ const addChainNode = async (models, req: Request, res: Response, next: NextFunct
...
       active: true,
       base: req.body.base,
+      website: req.body.website ? req.body.website : '',
+      discord: req.body.discord ? req.body.discord : '',
+      telegram: req.body.telegram ? req.body.telegram : '',
+      github: req.body.github ? req.body.github : '',
+      element: req.body.element ? req.body.element : '',
+      description: req.body.description ? req.body.description : '',
+      type: req.body.type ? req.body.type : 'chain',
     });
```
Comment (zakhap) [inline server/routes/addChainNode.ts:58]: Add line for `substrate_spec`

**Review by zakhap — CHANGES_REQUESTED** [review]
Just a few little notes, but very close to being merged in

File: client/scripts/app.ts (hunk 1 above)
Comment (maxwellfoley3) [reply client/scripts/app.ts:231]: With this line of code in, it's impossible to load a dynamically created Substrate chain, as it mandates that all chains must be explicitly numerated in the `ChainNetwork` list, aside from tokens. This check for undefined chains seems redundant with the check on line 305, so it should be fine to delete.

File: client/scripts/views/modals/create_community_modal.ts [hunk 4]
```diff
@@ -134,33 +139,212 @@ const CreateCommunityModal: m.Component<IAttrs, IState> = {
...
+          onChangeHandler: (v) => { vnode.state.substrate_spec = v; },
```
Comment (zakhap) [inline client/scripts/views/modals/create_community_modal.ts:222]: this validation needs to be more like the one in the `spec_settings.ts` file. And we should also add the "Test Button" to make sure that the spec and everything works (connects to chain via api). This should mostly be copy/pasta work.

**Review by zakhap — CHANGES_REQUESTED** [review]
Everything seems to work fine when given valid inputs. Given bad inputs (not properly formatted spec, but still valid json) it'll create a bad chainnode.

File: server/routes/addChainNode.ts [hunk 5]
```diff
@@ -23,6 +24,13 @@ const addChainNode = async (models, req: Request, res: Response, next: NextFunct
   if (!req.body.id || !req.body.name || !req.body.symbol || !req.body.network || !req.body.node_url || !req.body.base) {
     return next(new Error(Errors.MissingParams));
   }
+  if (req.body.substrate_spec) {
```
Comment (jnaviask) [inline server/routes/addChainNode.ts:27]: We should ideally do server-side validation of the spec, using the same logic from `/editSubstrateSpec`, and reject the chain if it doesn't work.

**Review by zakhap — APPROVED** [review]
This looks good to me! Just would like @jnaviask to look at the route check once before merging to master.

File: server/util/testSubstrateSpec.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    typesBundle: unpackedSpec['typesBundle'],
+    typesChain: unpackedSpec['typesChain'],
+    typesSpec: unpackedSpec['typesSpec'],
+  };
+
+  const provider = new WsProvider(constructSubstrateUrl(nodeUrl), false);
+  try {
+    await provider.connect();
+  } catch (err) {
+    return cb(new Error('failed to connect to node url'));
```
Comment (jnaviask) [inline server/util/testSubstrateSpec.ts:30]: We shouldn't be calling `next` here -- we should just throw an error and handle it in the route, then `return next` there.

**Review by jnaviask — CHANGES_REQUESTED** [review]
One small comment on route control flow, otherwise looks good to me.
